### PR TITLE
feature-intern-ResolveDependencyConflictsNew to main

### PR DESCRIPTION
**Pull request by NadiamB**
My code is currently behind an environment variable not yet set. Will work with @brettfo to put this behind the experiment flag.

Looking for feedback on implementation, unit tests, variable names, etc. 

DependencyConflictResolver.cs is the main code.

MSBuildHelper.cs and MSBuildHelperTests.cs  were used for tests

UpdateWorkerTests.Sdk.cs was lightly used with testing with Brett

Added temporary commit to make smoke tests pass.

Smoke Test PR: 
https://github.com/dependabot/smoke-tests/pull/219

This is the PR that adds the experiment:
https://github.com/github/dependabot-api/pull/5511

### CODE REVIEW

File: .vs/VSWorkspaceState.json
Comment (JamieMagee) [inline .vs/VSWorkspaceState.json:1]: I don't think the `.gitignore` got applied correctly.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs [hunk 1]
```diff
@@ -493,6 +497,792 @@ public async Task DependencyConflictsCanBeResolved()
         }
     }
 
+    // Scenario 1
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs:500]: nit: not sure how common this is to be used, but would love to see all of these "Scenario X" tests wrapped in a [`#region`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives#defining-regions)

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs [hunk 2]
```diff
@@ -493,6 +497,792 @@ public async Task DependencyConflictsCanBeResolved()
...
+            Assert.Equal("Azure.Core", resolvedDependencies[4].Name);
+            Assert.Equal("1.22.0", resolvedDependencies[4].Version);
+        }
+        finally
+        {
+            repoRoot.Delete(recursive: true);
+        }
+    }
+
+    // Scenario 10, Out of Scope
```
[the remark this answers is not in the document]
Comment (jpinz) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs:979]: My only other suggestion here is that I don't think the actual scenario # matters much here, so I might put this example at the end, it feels weird in the middle of all of the working ones.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 3]
```diff
@@ -24,6 +25,7 @@ internal static class SdkPackageUpdater
 
         // Get the set of all top-level dependencies in the current project
         var topLevelDependencies = MSBuildHelper.GetTopLevelPackageDependencyInfos(buildFiles).ToArray();
+        // TODO: Restore the code. It was commented out because it was being problematic and not reporting that some dependencies needed updating.
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:28]: q: this feels irrelevant now?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 4]
```diff
@@ -324,7 +327,21 @@ private static async Task AddTransitiveDependencyAsync(string repoRootPath, stri
...
+                    {
+                        var originalDependency = updatedTopLevelDependencies[dependencyIndex];
+                        updatedTopLevelDependencies[dependencyIndex] = originalDependency with { Version = previousDependencyVersion };
+                    }
+
+                }
+                Dependency[] update = [new Dependency(dependencyName, newDependencyVersion, DependencyType.PackageReference)];
+                Dependency[]? resolvedDependencies = await MSBuildHelper.ResolveDependencyConflicts(repoRootPath, projectFile.Path, tfm, updatedTopLevelDependencies, update, logger);
+
+                //Dependency[]? resolvedDependencies = await MSBuildHelper.ResolveDependencyConflicts(repoRootPath, projectFile.Path, tfm, updatedTopLevelDependencies, update, logger);
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:344]: nit: remove commented out code please

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs [hunk 5]
```diff
@@ -331,7 +334,187 @@ internal static async Task<bool> DependenciesAreCoherentAsync(string repoRoot, s
         }
     }
 
-    internal static async Task<Dependency[]?> ResolveDependencyConflicts(string repoRoot, string projectPath, string targetFramework, Dependency[] packages, Logger logger)
+    internal static bool UseNewDependencySolver()
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs:337]: nit: I think this method could just be simplified to `return Environment.GetEnvironmentVariable("UseNewNugetPackageResolver") == "true"`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs [hunk 6]
```diff
@@ -331,7 +334,187 @@ internal static async Task<bool> DependenciesAreCoherentAsync(string repoRoot, s
...
+            .Select(package => new PackageToUpdate
+            {
+                packageName = package.Name,
+                newVersion = package.Version.ToString()
+            })
+            .ToList();
+
+            foreach (var existing in existingPackages)
+            {
+                var x = packagesToUpdate.Where(p => p.packageName == existing.packageName);
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs:390]: nit: could you please use a more descriptive variable other than `x`?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs [hunk 7]
```diff
@@ -331,7 +334,187 @@ internal static async Task<bool> DependenciesAreCoherentAsync(string repoRoot, s
...
+                false,
+                false,
+                false
+            ))
+            .ToList();
+
+            // Return as array
+            Dependency[] candidatePackagesArray = candidatePackages.ToArray();
+
+            var TargetFrameworks = new NuGetFramework[] { NuGetFramework.Parse(targetFramework) };
```
Comment (jpinz) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs:479]: nit: could you use `targetFrameworks` instead?

File: .github/workflows/smoke.yml [hunk 8]
```diff
@@ -14,7 +14,7 @@ concurrency:
 
 env:
   GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-  SMOKE_TEST_BRANCH: main
+  SMOKE_TEST_BRANCH: feature-DependencySolver
```
[the remark this answers is not in the document]
Comment (brettfo) [reply .github/workflows/smoke.yml:17]: We added a smoke test, this was just to make sure it was green in CI.  The commit that added this will be removed prior to merging.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Cli.Test/EntryPointTests.Update.cs [hunk 9]
```diff
@@ -406,7 +406,11 @@ private static async Task Run(Func<string, string[]> getArgs, (string Path, stri
 
                 try
                 {
-                    await MockNuGetPackagesInDirectory(packages, path);
+                    if (packages != null)
```
[the remark this answers is not in the document]
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Cli.Test/EntryPointTests.Update.cs:409]: The method `MockNuGetPackagesInDirectory()` has a `null`-check built-in, so this can be reverted.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs (hunk 6 above)
Comment (brettfo) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs:390]: Haha, this one is my fault.  We were doing some pair coding and I suggested `x` just so we could get it working, then we'd give it a better name later.  I guess later is now.

Unrelated, but on this same line, NuGet package names are case insensitive; this should be `string.Equals(p.packageName, existing.packageName, StringComparison.OrdinalIgnoreCase)`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 10]
```diff
@@ -1,4 +1,5 @@
 using System.Collections.Immutable;
+using System.Reflection.Metadata;
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:2]: This can be deleted.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs [hunk 11]
```diff
@@ -0,0 +1,680 @@
...
+using NuGetUpdater.Core.Analyze;
+
+using Task = System.Threading.Tasks.Task;
+
+namespace NuGetUpdater.Core;
+
+// Data type to store information of a given package
+public class PackageToUpdate
+{
+    public string packageName { get; set; }
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs:22]: Make public members PascalCased, e.g., `public string PackageName { get; set; }`.  Same for other properties.

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs [hunk 12]
```diff
@@ -0,0 +1,680 @@
...
+            message.AppendLine($"  checking {source.Name}");
+            var sourceRepository = Repository.Factory.GetCoreV3(source);
+            var feed = await sourceRepository.GetResourceAsync<MetadataResource>(cancellationToken);
+            if (feed is null)
+            {
+                message.AppendLine($"    feed for {source.Name} was null");
+                continue;
+            }
+
+            var existsInFeed = await feed.Exists(
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs:95]: This call can throw if it's a private package source with bad authentication.  Wrap this section like is done [here](https://github.com/dependabot/dependabot-core/blob/4aac28c263db6b55c2c4ff326a57b49de034270e/nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Analyze/NuGetContext.cs#L107-L126).

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs [hunk 13]
```diff
@@ -324,7 +327,21 @@ private static async Task AddTransitiveDependencyAsync(string repoRootPath, stri
         {
             foreach (string tfm in targetFrameworks)
             {
-                Dependency[]? resolvedDependencies = await MSBuildHelper.ResolveDependencyConflicts(repoRootPath, projectFile.Path, tfm, updatedTopLevelDependencies, logger);
+                if (MSBuildHelper.UseNewDependencySolver())
+                {
+                    // Find the index of the dependency we are updating and revert it to the previous version
+                    int dependencyIndex = Array.FindIndex(updatedTopLevelDependencies, d => d.Name == dependencyName);
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Updater/SdkPackageUpdater.cs:333]: use `string.Equals(d.Name, dependencyName, StringComparison.OrdinalIgnoreCase)`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs [hunk 14]
```diff
@@ -0,0 +1,680 @@
...
+                // If there are dependencies
+                if (dependencyList != null)
+                {
+                    foreach (PackageToUpdate dependency in dependencyList)
+                    {
+                        // Check if the dependency is in the existing packages
+                        foreach (PackageToUpdate existingPackage in existingPackages)
+                        {
+                            // If you find the dependency
+                            if (dependency.packageName == existingPackage.packageName)
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs:371]: Use `string.Equals(dependency.packageName, existingPackage.packageName, StringComparison.OrdinalIgnoreCase)`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs [hunk 15]
```diff
@@ -0,0 +1,680 @@
...
+    public async Task<bool> IsCompatibleAsync(PackageToUpdate parent, PackageToUpdate child, string targetFramework, string projectDirectory)
+    {
+        // Get the dependencies of the parent
+        List<PackageToUpdate> dependencies = await GetDependenciesAsync(parent, targetFramework, projectDirectory);
+
+        foreach (PackageToUpdate dependency in dependencies)
+        {
+
+            // If the child is present
+            if (dependency.packageName == child.packageName)
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/DependencyConflictResolver.cs:515]: Use `string.Equals(dependency.packageName, child.packageName, StringComparison.OrdinalIgnoreCase)`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs [hunk 16]
```diff
@@ -331,7 +334,187 @@ internal static async Task<bool> DependenciesAreCoherentAsync(string repoRoot, s
...
+            // Create a duplicate set of existingPackages for flexible package reference addition and removal 
+            List<PackageToUpdate> existingDuplicate = new List<PackageToUpdate>(existingPackages);
+
+            // Bool to keep track of if anything was added to the existingDuplicate list
+            bool added = false;
+
+            // If package 'isnt there, add it to the existingDuplicate list
+            foreach (PackageToUpdate package in packagesToUpdate)
+            {
+                if (!existingDuplicate.Any(p => p.packageName == package.packageName))
```
Comment (brettfo) [inline nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core/Utilities/MSBuildHelper.cs:406]: Use `string.Equals(p.packageName, package.packageName, StringComparison.OrdinalIgnoreCase)`

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs [hunk 17]
```diff
@@ -479,7 +479,11 @@ public async Task DependencyConflictsCanBeResolved()
                 new Dependency("Some.Package", "1.2.0", DependencyType.PackageReference),
                 new Dependency("Some.Other.Package", "1.0.0", DependencyType.PackageReference),
             };
-            var resolvedDependencies = await MSBuildHelper.ResolveDependencyConflicts(repoRoot.FullName, projectPath, "net8.0", dependencies, new Logger(true));
+            var update = new[]
+            {
+                new Dependency("Some.Other.Package", "1.2.0", DependencyType.PackageReference),
+            };
+            var resolvedDependencies = await MSBuildHelper.ResolveDependencyConflicts(repoRoot.FullName, projectPath, "net8.0", dependencies, update, new Logger(true));
```
[the remark this answers is not in the document]
Comment (NadiamB) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs:486]: Followed the same format as the rest of the file since they also pass it net8.0 directly in

**Review by brettfo — APPROVED** [review]

**Review by abdulapopoola — APPROVED** [review]

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs (hunk 1 above)
Comment (jpinz) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs]: Looks like you did it, but could you name the region as well? Maybe like `NugetDependencyResolutionTests` or something?

File: nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs [hunk 18]
```diff
@@ -493,6 +497,786 @@ public async Task DependencyConflictsCanBeResolved()
         }
     }
 
+    #region
```
[the remark this answers is not in the document]
Comment (NadiamB) [reply nuget/helpers/lib/NuGetUpdater/NuGetUpdater.Core.Test/Utilities/MSBuildHelperTests.cs:500]: Will do these in follow up PR

File: .github/workflows/smoke.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (kbukum1) [reply .github/workflows/smoke.yml:17]: We just revert back this change because we are getting error.

[the remark this answers is not in the document]
Comment (kbukum1) [reply .github/workflows/smoke.yml:17]: https://github.com/dependabot/dependabot-core/pull/10437
